server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -711,4 +792,42 @@ void sendFiles(Store store, StoreFileMetaData[] files, Function<StoreFileMetaDat
     protected void failEngine(IOException cause) {
         shard.failShard("recovery", cause);
     }
+
+    private static final class StepListener<V> implements ActionListener<V> {
```

Comment:
I like this class. I think we should make it a primitive and test / document it.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/37327.